### PR TITLE
feat(analytics): cross-tier span propagation (client fetch header + server withSpan({ request }))

### PR DESCRIPTION
**Pull request by mantrakp04**
Stacked on #1719 (`feat/custom-events-spans-serverless`).

## What
Auto-links a customer's **backend** span to the caller's **client** session with **zero glue**. When a browser calls the customer's own backend, a server span opened with `serverApp.withSpan(type, { request }, fn)` parents under that tab's `$refresh-token → $session-replay → $session-replay-segment` chain — the same tree browser events already attach to.

```ts
// client: nothing to do — analytics on ⇒ same-origin fetches carry the header

// server: only where you want a span
export const POST = (req: Request) =>
  stackServerApp.withSpan("checkout", { request: req }, async (span) => {
    span.trackEvent("checkout.validated");   // parent_span_ids = [rti-…, sri-…, srsi-…]
  });
```

## How
One header, `x-hexclave-span-context: v1.<base64url-json>`, carrying **raw** ids (`{ projectId, sessionReplaySegmentId?, sessionReplayId?, customParentSpanIds? }`). The client knows only the segment id; the **server** resolves `rti-` from the session; the **backend** derives `sri-` and composes all the `rti-`/`sri-`/`srsi-`/`cs-` prefixes. Nothing client-side ever fabricates a system id.

- **`span-propagation.ts`** — versioned header codec, a same-origin `fetch` auto-wrapper (idempotent via a global marker, preserves native header precedence then adds ours, skips `no-cors`, restores the exact original on uninstall), and the origin predicate.
- **`server-request-context.ts`** — AsyncLocalStorage ambient request context, so everything inside a `withSpan({ request })` callback inherits it (even bare `serverApp.trackEvent`).
- **`server-app-impl`** — `withSpan`/`trackEvent({ request })` resolve the context; the server-telemetry buffer is re-keyed from `userId` → the full `(user, refreshToken, replay, segment)` context so two requests never share a batch's ancestry.
- **backend `events/batch`** — accepts the forwarded `refresh_token_id`/`session_replay_id` on **server auth** (the secret key is trusted), **rejects** them on client auth, composes `[rti-, sri-, srsi-, cs-]`, and stamps the scalar columns (not just `parent_span_ids`).
- **`client-app-impl`** — installs the wrapper on init behind `analytics.spanPropagation` (same-origin default + exact-origin `targets` allowlist); reads the **live** per-tab id so it tracks sign-out rotation.

## Trust / safety
Best-effort telemetry, never authz. Only `rti-`/`userId`/`project`/`branch` are server-trusted; `sri-`/`srsi-`/custom parents are **untrusted labels** from a client-controlled header. Header is **same-origin only** by default (cross-origin would leak context + trip CORS preflight), version/size/uuid/project validated, and any bad header is ignored — never thrown into the request.

Design validated with Codex (gpt-5.5 xhigh): versioned opaque header (deliberately **not** `traceparent`, to avoid clobbering the customer's own OTel), `Date.now()`-free RMT semantics, and the same-origin propagation default.

## Tests
- **24** unit tests (`span-propagation.test.ts`): codec round-trip / rejection / caps, same-origin predicate, and the fetch wrapper's edge cases (Request vs string vs URL input, init-headers precedence, no-cors, idempotency, clean uninstall).
- **128** existing template tests still green (server-buffer refactor caused no regression).
- **2** e2e (`analytics-events-batch.test.ts`): a server-auth batch with forwarded context → CH event carries `[rti-,sri-,srsi-,cs-]` + stamped scalar columns; client auth rejects the forwarded context.
- Typecheck green across backend, template, and all four generated SDKs (js/react/next/tanstack-start); changed-file eslint clean.

## Follow-up (separate stack)
Framework **adapters** — tRPC, Convex, oRPC, ElysiaJS — build on this primitive so consumers never pass `{ request }` themselves (the adapter pulls it from the framework's context). They're a full auth+telemetry integration, shipping on their own stacked branch.

---
## Summary by cubic
Links backend spans/events to the caller’s browser session and trace via same-origin `traceparent` and `x-hexclave-span-context`, plus request-aware server APIs. Previously backend spans were isolated; now they join the client’s trace and session chain with strict parsing shared across SDKs and backend.

- Client: auto-adds `traceparent` and `x-hexclave-span-context` on same-origin `fetch` (allowlist via `analytics.spanPropagation.targets`, skips `no-cors`); never overwrites an explicit header.
- Client APIs: `getSpanPropagationHeaders({ parentIds?, root? })` for non-`fetch`; span handle kit (`span.withSpan`, `span.run`, `span.getPropagationHeaders`, `span.fetch`) with exact-only parenting via prologue-open sync frames, propagating global and nested parents.
- Server: `withSpan(..., { request })` and `trackEvent(..., { request })` join the incoming W3C trace and merge client-propagated labels/parents before server frames; ambient via AsyncLocalStorage; detaches when an explicit user differs; surfaces request telemetry failures.
- Backend: `customer-request-observability` mints a backend request span linked to the caller’s trace once tenancy is proven; `events/batch` accepts forwarded `refresh_token_id`/`session_replay_id` on server auth (rejects on client), composes `[rti-, sri-, srsi-, cs-]`, stamps scalar columns; `$token-refresh` resolves request observability early.
- Shared: propagation header codec moved to `@hexclave/shared` for one-wire, strict parsing across SDKs and backend.

- Migration
  - On by default; disable via `analytics.spanPropagation.enabled = false`.
  - Pass `{ request }` to `withSpan` or `trackEvent` to link server spans/events.
  - For split frontend/API domains, add the API origin to `analytics.spanPropagation.targets` and allow headers in CORS: `Access-Control-Allow-Headers: x-hexclave-span-context, traceparent`.

<sup>Written for commit 5f1463a85662bbb994fc6223d474d2f23c0fa41a. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/hexclave/hexclave/pull/1726?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added cross-tier span propagation for analytics, improving end-to-end request tracing.
  * Server-side event tracking can now automatically associate requests with the right session context.
  * Telemetry batches now support additional session identifiers for more accurate attribution.

* **Bug Fixes**
  * Tightened validation so client-side requests can’t submit server-derived session fields.
  * Improved session and parent-span linking for forwarded telemetry, reducing mismatched event ancestry.

---

## Update — propagation moved onto the shared wire contract

- **New `packages/shared/src/utils/span-context-codec.tsx`** — the propagation header's encode/decode now lives in shared, so all three readers parse it identically: the client SDK that writes it, `server-request-context` on the receiving side, and Hexclave's own backend when it resolves the tenancy for a cross-tier request.
- `span-propagation.ts` and `server-request-context.ts` build parent paths from that codec plus the `analytics-wire` id helpers (uuid ⇄ W3C), instead of hand-rolled parsing.

<sub>Slice: 5 files. Part of a stack-wide split of one working tree into the PR each change belongs to.</sub>

### CODE REVIEW

**Comment by mantrakp04** [comment]
@coderabbitai review

File: packages/template/src/lib/hexclave-app/apps/implementations/client-app-impl.ts [hunk 1]
```diff
@@ -748,6 +750,35 @@ export class _HexclaveClientAppImplIncomplete<HasTokenStore extends boolean, Pro
...
+    // which reflects sign-out rotation). Same-origin only unless extra exact origins
+    // are allowlisted. Idempotent — installFetchSpanPropagation no-ops if already on.
+    if (
+      analyticsEnabled
+      && isBrowserLike()
+      && this._eventTracker
+      && this._analyticsOptions?.spanPropagation?.enabled !== false
+    ) {
+      const eventTracker = this._eventTracker;
+      const allowedOrigins = this._analyticsOptions?.spanPropagation?.targets ?? [];
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply packages/template/src/lib/hexclave-app/apps/implementations/client-app-impl.ts:880]: Fixed in b59bab84c.

The review was valid: the browser fetch wrapper is process-global, so the first app instance could permanently own the provider used to build `x-hexclave-span-context`. I changed the global marker from a boolean into a small shared state object. The wrapper still installs only once, but later app construction updates that shared provider, so fetches use the current app's project/session context and target policy instead of the first app's closed-over values.

I also added a regression test in `span-propagation.test.ts` that installs propagation twice and verifies the second app's context is what gets attached.

File: packages/template/src/lib/hexclave-app/apps/implementations/client-app-impl.ts [hunk 2]
```diff
@@ -748,6 +749,25 @@ export class _HexclaveClientAppImplIncomplete<HasTokenStore extends boolean, Pro
...
+    // which reflects sign-out rotation). Same-origin only unless extra exact origins
+    // are allowlisted. Idempotent — installFetchSpanPropagation no-ops if already on.
+    if (
+      analyticsEnabled
+      && isBrowserLike()
+      && this._eventTracker
+      && this._analyticsOptions?.spanPropagation?.enabled !== false
+    ) {
+      const allowedOrigins = this._analyticsOptions?.spanPropagation?.targets ?? [];
+      installFetchSpanPropagation({
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply packages/template/src/lib/hexclave-app/apps/implementations/client-app-impl.ts:881]: Fixed in b59bab84c.

This was the same root cause as the Greptile thread: `fetch` can only be wrapped globally, but the data used by the wrapper is per app instance. The fix keeps one wrapper and moves the per-app context into mutable shared state, so subsequent `StackClientApp` instances update the context source instead of being ignored.

The regression test covers the multi-app case directly: install app A, install app B, fetch, and assert the propagated header belongs to app B.

File: apps/backend/src/app/api/latest/analytics/events/batch/route.tsx [hunk 3]
```diff
@@ -203,8 +213,13 @@ export const POST = createSmartRouteHandler({
       }
     }
 
-    const recentSession = refreshTokenId == null ? null : await findRecentSessionReplay(prisma, { tenancyId, refreshTokenId });
-    const sessionReplayId = recentSession?.id ?? null;
+    // Prefer an explicitly forwarded replay id (server SDK — exact join); otherwise
+    // derive the caller's current rolling replay from their refresh token.
+    let sessionReplayId = body.session_replay_id ?? null;
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply apps/backend/src/app/api/latest/analytics/events/batch/route.tsx]: Fixed in b59bab84c.

The review was valid. Server/admin batches are trusted to identify the caller, but a forwarded `session_replay_id` still needs to be bound to the same project session root before we stamp it into analytics rows. The route now rejects `session_replay_id` unless there is a `refresh_token_id`, then verifies that the replay row belongs to this tenancy, that refresh token, and the user when a user is known.

That means the backend no longer accepts arbitrary UUID-shaped replay IDs and composes `rti-` / `sri-` / `srsi-` ancestry from unrelated inputs.

File: apps/backend/src/app/api/latest/analytics/events/batch/route.tsx [hunk 4]
```diff
@@ -203,8 +213,13 @@ export const POST = createSmartRouteHandler({
...
+    if (sessionReplayId == null && refreshTokenId != null) {
+      const recentSession = await findRecentSessionReplay(prisma, { tenancyId, refreshTokenId });
+      sessionReplayId = recentSession?.id ?? null;
+    }
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply apps/backend/src/app/api/latest/analytics/events/batch/route.tsx]: Fixed in b59bab84c.

This was valid and is covered by the same route-level guard. A server-auth batch can no longer provide `session_replay_id` without the refresh-token root: the endpoint now returns a 400 for that shape. If a replay id is provided with a refresh token, the replay must match that refresh token in the project tenancy before any replay parent is composed.

I added e2e coverage for both cases: valid server-auth replay context now uses a real replay row, and the missing/wrong refresh-token cases are rejected.

File: packages/template/src/lib/hexclave-app/apps/implementations/server-request-context.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+/** Runs `fn` with `context` ambient. Awaits the ALS load first so the very first
+ * call gets per-request isolation rather than racing the module load. */
+export async function runWithServerRequestContext<T>(context: ServerRequestSpanContext, fn: () => Promise<T>): Promise<T> {
+  await ensureAsyncContext();
+  if (als) {
+    return await als.run(context, fn);
+  }
+  const previous = syncCurrent;
+  syncCurrent = context;
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply packages/template/src/lib/hexclave-app/apps/implementations/server-request-context.ts]: Fixed in b59bab84c.

The fallback was validly risky: without `AsyncLocalStorage`, a single mutable request context can cross-contaminate concurrent server requests. I changed `server-request-context.ts` to fail closed in that runtime. If exact async context is unavailable, `runWithServerRequestContext` still runs the callback, but `getServerRequestContext()` returns null instead of exposing potentially wrong user/session ancestry.

That preserves telemetry recording while avoiding incorrect attribution.

File: packages/template/src/lib/hexclave-app/apps/implementations/server-app-impl.ts [hunk 6]
```diff
@@ -1890,53 +1995,89 @@ export class _HexclaveServerAppImplIncomplete<HasTokenStore extends boolean, Pro
...
+      // no-cors requests are still skipped.
+      fetch: (input: RequestInfo | URL, init?: RequestInit) => {
+        try {
+          const initWithHeader = buildFetchInitWithSpanContext({
+            input,
+            init,
+            headerValue: span.getPropagationHeaders()[SPAN_CONTEXT_HEADER],
+            selfOrigin: null,
+            allowedOrigins: [],
+            bypassOriginPolicy: true,
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply packages/template/src/lib/hexclave-app/apps/implementations/server-app-impl.ts]: Fixed in b59bab84c.

Valid security concern. Server runtimes do not have a reliable browser-like current origin, so treating `span.fetch()` as explicit permission to attach `x-hexclave-span-context` to any URL was too broad. I removed the `bypassOriginPolicy` path for server `span.fetch()`, so it now fails closed instead of leaking project/session/span identifiers to third-party services.

For trusted server-to-server calls where the developer controls the target, `span.getPropagationHeaders()` remains the explicit escape hatch.

File: packages/template/src/lib/hexclave-app/apps/implementations/server-app-impl.ts [hunk 7]
```diff
@@ -1890,53 +1995,89 @@ export class _HexclaveServerAppImplIncomplete<HasTokenStore extends boolean, Pro
         this._trackServerEvent(eventType, data, { ...trackOptions, parentIds: [span, ...trackOptions?.parentIds ?? []] }, userId),
       startSpan: (childType: string, childOptions?: StartSpanOptions) =>
         this._startServerSpan(childType, { ...childOptions, parentIds: [span, ...childOptions?.parentIds ?? []] }, userId),
+      withSpan: <T,>(childType: string, optionsOrFn: StartSpanOptions | ((child: Span) => Promise<T> | T), maybeFn?: (child: Span) => Promise<T> | T) =>
+        withSpanImpl((type, opts) => span.startSpan(type, opts), childType, optionsOrFn, maybeFn),
+      run: <T,>(fn: () => T) => runWithSpanFrame(span.ref(), fn),
+      // Pinned to exactly this span's frozen chain; carries the caller's segment
+      // identity when this span was resolved from a request. Raw ids — the
+      // receiving backend applies the prefixes.
+      getPropagationHeaders: () => ({
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply packages/template/src/lib/hexclave-app/apps/implementations/server-app-impl.ts:2019]: Fixed in b59bab84c.

The propagation payload was missing part of the replay ancestry. Server spans created inside `withSpan({ request })` can carry both `sessionReplayId` and `sessionReplaySegmentId` in their batch context, but `getPropagationHeaders()` only forwarded the segment. I now include `sessionReplayId` when present, so downstream services can reconstruct the full refresh-token -> replay -> segment chain instead of only the segment leaf.

File: packages/template/src/lib/hexclave-app/apps/implementations/span-context.ts [hunk 8]
```diff
@@ -54,32 +68,99 @@ async function ensureAsyncContext(): Promise<void> {
...
+/**
+ * Whether an EXACT async-context primitive backs the ambient frames right now
+ * (AsyncLocalStorage today; TC39 AsyncContext when browsers ship it). When
+ * false, getAmbientSpanRefs() serves the shared sync stack, which can mix
+ * frames from concurrently interleaved async flows — readers use this to apply
+ * the `ambientParenting` policy ("exact" drops the frames, "best-effort" keeps
+ * them). Frames are always RECORDED either way; only reading is gated, so the
+ * policy can differ per consumer without losing context.
+ */
+export function isExactAsyncContextActive(): boolean {
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply packages/template/src/lib/hexclave-app/apps/implementations/span-context.ts]: Fixed in b59bab84c.

This helper was indeed unused after the surrounding code changed. I removed `isExactAsyncContextActive` rather than keeping a misleading exported policy hook. The actual policy is now expressed at the reader call sites through `getAmbientSpanRefs({ includeSuspendedSyncFrames })`.

File: packages/template/src/lib/hexclave-app/apps/implementations/span-context.ts [hunk 9]
```diff
@@ -54,32 +68,99 @@ async function ensureAsyncContext(): Promise<void> {
...
+  const syncFrame: SyncFrame = { ref, prologueOpen: true };
+  syncStack.push(syncFrame);
+  const pop = () => {
+    const index = syncStack.lastIndexOf(syncFrame);
     if (index !== -1) syncStack.splice(index, 1);
+  };
+  try {
+    const result = fn();
+    syncFrame.prologueOpen = false;
+    if (result instanceof Promise) {
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply packages/template/src/lib/hexclave-app/apps/implementations/span-context.ts]: Fixed in b59bab84c.

Valid edge case. `span.run()` previously used `instanceof Promise`, which only recognizes same-realm native promises. I changed it to detect promise-like return values and schedule cleanup through `Promise.resolve(result).finally(...)`, so thenables and cross-realm promises keep the best-effort fallback frame visible until they settle.

There is a regression test in `span-context.test.ts` using a custom thenable to pin this behavior.

**Comment by mantrakp04** [comment]
@greptile-ai review

File: packages/template/src/lib/hexclave-app/apps/implementations/server-app-impl.ts [hunk 10]
```diff
@@ -1723,22 +1725,101 @@ export class _HexclaveServerAppImplIncomplete<HasTokenStore extends boolean, Pro
...
+
+  /**
+   * The batch context an item is buffered/attributed under: the ambient request
+   * context when inside a `{ request }` scope, else just the explicit userId. A
+   * later explicit userId always wins over the request-derived one.
+   */
+  private _currentServerBatchContext(explicitUserId: string | null): ServerRequestSpanContext {
+    const ambient = getServerRequestContext();
+    if (ambient) {
+      return { ...ambient, userId: explicitUserId ?? ambient.userId };
```
[the remark this answers is not in the document]
Comment (mantrakp04) [reply packages/template/src/lib/hexclave-app/apps/implementations/server-app-impl.ts]: Fixed in 8323226cc. Explicit server user attribution now keeps request-derived refresh/replay/segment/custom-parent context only when it exactly matches the authenticated request user. A different explicit user, or an explicit user on anonymous request context, gets a fully detached batch context. This is applied both while resolving `{ request }` and for calls nested inside ambient request spans. Added focused regressions for implicit/matching, mismatched, and anonymous identities; template lint/typecheck and JS/React/Next/TanStack typechecks pass. The live backend accepted the properly detached cross-user server batch; its later local analytics-query assertion was unavailable (404), while the dangerous mixed replay payload remains rejected by the backend.

**Comment by mantrakp04** [comment]
Superseded by the consolidated [observability base PR](https://github.com/hexclave/hexclave/pull/1976). The original PR, commits, and review history are retained.
